### ISSUE TITLE
Horovod MPI Tensorflow benchmark

### ISSUE THREAD

**Issue by nadeengebara** [horovod/horovod#1066]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet):  Tensorflow
2. Framework version: 1.13
3. Horovod version: 0.16-1
4. MPI version: 1.10.2
5. CUDA version:-
6. NCCL version:
7. Python version:2.7.12
8. OS and version: Linux ubuntu 16.0.4

Hi, 
 I've modified the mpi_operations.cc in horovod/common/ops to display the gradients and their aggregated values upon calling horovod's MPIAllReduce::Execute.
Then i ran the tensorflow synthetic benchmark using horovod as follows:

HOROVOD_TIMELINE= ../timeline.json horovodrun -np 2 -H ...  python tf_cnn_benchmark.py --num_warmup_batches 0 --num_batches 1 --variable_update horovod

I noticed that the horovod allreduce was actually executing twice where the first time entries .size was <=1 and the second time entries.size>1 resulting in copying data to the fusion buffer. 
Can you please clarify why this is happening as can be seen in the attached image.

Thank you!

<img width="894" alt="timeline" src="https://user-images.githubusercontent.com/17450628/57572483-e9aa5b80-7412-11e9-8391-40dfd2cbcd4d.PNG">

**Comment by alsrgv** [comment]
@nadeengebara, fusion buffering is done based on timing & size of the gradients.  Please see https://github.com/horovod/horovod/blob/master/docs/tensor-fusion.md for more details.

**Comment by nadeengebara** [comment]
I see, thanks !

**Comment by nadeengebara** [comment]
Hi Alex,
 Hope all is well. I just have a quick question. When running Horovod with the tensorflow benchmark and a single GPU on each machine in a multi-node setup, it appeared that MPI and not NCCL was used for inter-node communication.
Does horovod only use NCCL when GPUs belong to a single machine, and falls back on MPI for inter-node communication, or is that as a result of the way the variable update is defined in the tensorflow  benchmark.
As usual, thank you very much for your constant helpful responses.

Best,
Naden

**Comment by alsrgv** [comment]
@nadeengebara, that's odd. There is a mode where MPI is used across nodes and NCCL is used within called hierarchical allreduce.  It's not enabled by default.  By default, Horovod uses NCCL both within the server and across servers.
